src/core/net/dns_client.cpp

```diff
@@ -1106,6 +1186,132 @@ void Client::HandleTimer(void)
     }
 }
 
+#if OPENTHREAD_CONFIG_DNS_OVER_TCP_ENABLE
+void Client::HandleTcpTimer(void)
+{
+    // If message wasn't received in timeout time close connection.
+    mEndpoint.SendEndOfStream();
+    mTcpInitialized = false;
+}
+
+void Client::HandleTcpEstablished(otTcpEndpoint *aEndpoint)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    mEndpoint.SendByReference(mSendLink, 0);
+}
+
+void Client::HandleTcpEstablishedCallback(otTcpEndpoint *aEndpoint)
+{
+    LogDebg("Dns tcp connection established.\n");
```

Comment:
What Abtin wrote about managing TCP buffers is correct.

Regarding `otMessage`, the TCP stack does not use them because it could lead to deadlock under low-memory situations (e.g,. TCP cannot allocate a packet to send out data because too many `otMessage`s are tied up in its send queue).

I recognize that the `otLinkedBuffer`s take more effort to use. To help alleviate this, we provide a circular buffer library (see `tcp_ext.h`) that does much of that management automatically, and provides the user of TCP with a fixed-size buffer abstraction for the send buffer. Perhaps that could be useful?

I'll try to take a closer look at the code today to give more specific guidance.

Reply:
fixed logging.